engine.py

```diff
@@ -23,200 +19,240 @@
         local_python_path = os.path.abspath(os.path.join( os.path.dirname(__file__), "python"))
         os.environ["TANK_NUKE_ENGINE_MOD_PATH"] = local_python_path
-        
-    
+
+    @property
+    def has_ui(self):
+        """
+        Whether Nuke is running a GUI session.
+        """
+        return self._ui_enabled
+
+    @property
+    def hiero(self):
+        """
+        Whether Nuke is running in Hiero mode.
+        """
+        return self._hiero
+
     def pre_app_init(self):
         """
-        Called at startup, but after QT has been initialized
+        Called at startup, but after QT has been initialized.
         """
-        # note! not using the import as this confuses nuke's calback system
+        if self.hiero:
+            return
+
+        # Note! not using the import as this confuses nuke's calback system
         # (several of the key scene callbacks are in the main init file...)
         import tk_nuke
         
-        # make sure callbacks tracking the context switching are active
+        # Make sure callbacks tracking the context switching are active.
         tk_nuke.tank_ensure_callbacks_registered()
-        
-       
+
     def post_app_init(self):
         """
-        Called when all apps have initialized
+        Called when all apps have initialized.
         """
-        # render the menu!
-        if self._ui_enabled:
-            
-            # note! not using the import as this confuses nuke's calback system
+        # Render the menu!
+        if self.has_ui:
+            # Note! not using the import as this confuses nuke's calback system
             # (several of the key scene callbacks are in the main init file...)            
             import tk_nuke
 
             menu_name = "Shotgun"
             if self.get_setting("use_sgtk_as_menu_name", False):
                 menu_name = "Sgtk"
-            
-            # create menu
+
             self._menu_generator = tk_nuke.MenuGenerator(self, menu_name)
             self._menu_generator.create_menu()
-            
-            # initialize favourite dirs in the file open/file save dialogs
-            self.__setup_favorite_dirs()
-            
-            # register all panels with nuke's callback system
-            # this will be used at nuke startup in order
-            # for nuke to be able to restore panels 
-            # automatically. For all panels that exist as
-            # part of saved layouts, nuke will look through
-            # a global list of registered panels, try to locate
-            # the one it needs and then run the callback         
-            for (panel_id, panel_dict) in self.panels.iteritems():
-                nukescripts.panels.registerPanel(panel_id, panel_dict["callback"])
-            
-        # iterate over all apps, if there is a gizmo folder, add it to nuke path
+
+            if self.hiero:
+                import hiero
+                def set_project_root(event):
```

Comment:
i think we should avoid defining methods like this if we can - sometimes we do, but it's more if it's a tiny little helper that is only of local relevance - almost like a lambda. This doesn't feel like that - seems like it should be a proper part of the engine?

Also, maybe there should be two methods `init_hiero` and `init_nuke` that this method calls?. That way maybe we could avoid the return inside the hiero if clause too?